File: Algorithms/084_LargestRectangleInHistogram.py
class Solution(object):
    def largestRectangleArea(self, heights):
        """
        :type heights: List[int]
        :rtype: int
        """
        if len(heights) < 1:
            return 0
        h, indexes = [heights[0]], [0]
        largestArea = heights[0]
        i = 1
        while i < len(heights):
            if heights[i] >= h[-1]:
                h.append(heights[i])
                indexes.append(i)
            else:
                j = 0
                while h and heights[i] < h[-1]:
                    j = indexes[-1]
                    largestArea = max(h[-1] * (i - j), largestArea)
                    h.pop(-1)
                    indexes.pop(-1)
                h.append(heights[i])
                indexes.append(j)
            i += 1
        while h:
            largestArea = max(h[-1] * (i - indexes[-1]), largestArea)
            h.pop(-1)
            indexes.pop(-1)
        return largestArea

File: Algorithms/cpp/084.LargestRectangleInHistogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& height) {
        if (height.size() == 0) {
            return 0;
        }
        stack<int> heights, indexes;
        heights.push(height[0]);
        indexes.push(0);
        int largestArea = height[0];
        int i = 1;
        while (i < height.size()) {
            if (height[i] >= heights.top()) {
                heights.push(height[i]);
                indexes.push(i);
            } else {
                int h, j;
                while (!heights.empty() && height[i] < heights.top()) {
                    h = heights.top();
                    j = indexes.top();
                    int area = h * (i - j);
                    if (area > largestArea) {
                        largestArea = area;
                    }
                    heights.pop();
                    indexes.pop();
                }
                heights.push(height[i]);
                indexes.push(j);
            }
            i++;
        }
        while (!heights.empty()) {
            if (heights.top() * (i - indexes.top()) > largestArea) {
                largestArea = heights.top() * (i - indexes.top());
            }
            heights.pop();
            indexes.pop();
        }
        return largestArea;
    }
};
